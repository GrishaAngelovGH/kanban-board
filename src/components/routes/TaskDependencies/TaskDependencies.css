.task-dependencies {
  height: calc(100vh - 145px);
}

@media only screen and (max-width: 576px) {
  .task-dependencies {
    height: calc(100vh - 200px);
  }
}