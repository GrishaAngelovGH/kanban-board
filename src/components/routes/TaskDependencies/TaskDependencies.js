import { Link } from "react-router-dom"

import Badge from "react-bootstrap/Badge"
import Button from "react-bootstrap/Button"
import ListGroup from "react-bootstrap/ListGroup"

import boardRepository from "persistent/persistentKanbanBoardRepository"

import useBackgroundImage from "hooks/useBackgroundImage"

import "./TaskDependencies.css"

const TaskDependencies = ({ taskId, columnId, showToastWithMessage }) => {
  const backgroundImage = useBackgroundImage()
  const columns = boardRepository.getColumns()
  const column = columns.find(v => v.id === columnId)
  const task = boardRepository.getTaskById(column, taskId)

  const dependencyTasks = boardRepository.getDependencyTasks(taskId, columnId)

  return (
    <div className="row g-0 vh-100" style={{ backgroundImage: backgroundImage ? `url(${backgroundImage})` : "none", backgroundSize: "cover" }}>
      <div className="col-12">
        <div className="row g-0 bg-light p-2 shadow align-items-center">
          <div className="col-2">
            <Link to="/" className="btn btn-light bi bi-arrow-left border"></Link>
          </div>
          <div className="col-8">
            <h3 className="m-0 text-center">Task Dependencies</h3>
          </div>
        </div>

        <div className="row g-0 mt-3 justify-content-center">
          <div className="col-10 col-md-6 bg-light rounded">
            <h3 className="m-0 text-center">{task.title}</h3>
          </div>
        </div>

        <div className="row g-0 mt-3 justify-content-around overflow-auto task-dependencies">
          <div className="col-10 col-md-4">
            <ListGroup>
              <ListGroup.Item className="bg-secondary-subtle text-secondary fw-bold text-center">
                All columns with tasks
              </ListGroup.Item>
              {
                columns
                  .filter(col => !col.markedAsDone && !col.items.every(v => boardRepository.isDependencyTask(v.id)))
                  .map(col => (
                    <ListGroup.Item key={col.id}>
                      <Badge className="mb-3">{col.title}</Badge>
                      {
                        col.items.map(v => (
                          <div key={v.id}>
                            {
                              !boardRepository.isDependencyTask(v.id) && (
                                <ListGroup.Item className="border border-3 rounded mb-3">
                                  <p>{v.title}</p>
                                  <Button
                                    size="sm"
                                    variant="outline-primary"
                                    className="w-100"
                                    disabled={v.id === taskId}
                                    onClick={(() => {
                                      boardRepository.addDependencyTask(taskId, columnId, v.id, col.id)
                                      showToastWithMessage("The dependency task is successfully added")
                                    })}
                                  >
                                    Add as a dependency
                                  </Button>
                                </ListGroup.Item>
                              )
                            }
                          </div>
                        ))
                      }
                    </ListGroup.Item>
                  ))
              }
            </ListGroup>
          </div>
          <div className="col-10 col-md-4">
            <ListGroup className="mt-3 mt-md-0">
              <ListGroup.Item className="bg-secondary-subtle text-secondary fw-bold text-center">
                Dependency tasks
              </ListGroup.Item>
              {
                dependencyTasks.map(v => (
                  <ListGroup.Item key={v.id}>
                    <p>{v.title}</p>
                    <Button
                      size="sm"
                      variant="outline-danger"
                      className="w-100"
                      onClick={(() => {
                        boardRepository.removeDependencyTask(taskId, columnId, v.id)
                        showToastWithMessage("The dependency task is successfully removed")
                      })}
                    >
                      Remove as a dependency
                    </Button>
                  </ListGroup.Item>
                ))
              }
            </ListGroup>
          </div>
        </div>
      </div>
    </div>
  )
}

export default TaskDependencies