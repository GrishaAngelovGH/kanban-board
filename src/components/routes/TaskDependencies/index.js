import TaskDependencies from "./TaskDependencies"

export default TaskDependencies