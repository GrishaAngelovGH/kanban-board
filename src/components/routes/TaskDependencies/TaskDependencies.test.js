import { render } from "@testing-library/react"

import {
  BrowserRouter as Router,
  Routes,
  Route
} from "react-router-dom"

import TaskDependencies from "./TaskDependencies"

jest.mock('persistent/persistentKanbanBoardRepository')

test("should render TaskDependencies route component", () => {
  const view = render(
    <Router>
      <Routes>
        <Route path="/" element={<TaskDependencies taskId="3792131195588513" />} />
      </Routes>
    </Router>
  )

  expect(view).toMatchSnapshot()
})