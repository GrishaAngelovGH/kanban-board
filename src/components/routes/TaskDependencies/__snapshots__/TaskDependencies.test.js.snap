// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TaskDependencies route component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row g-0 vh-100"
        style="background-image: url(geometric-triangle-shapes-background.jpg); background-size: cover;"
      >
        <div
          class="col-12"
        >
          <div
            class="row g-0 bg-light p-2 shadow align-items-center"
          >
            <div
              class="col-2"
            >
              <a
                class="btn btn-light bi bi-arrow-left border"
                href="/"
              />
            </div>
            <div
              class="col-8"
            >
              <h3
                class="m-0 text-center"
              >
                Task Dependencies
              </h3>
            </div>
          </div>
          <div
            class="row g-0 mt-3 justify-content-center"
          >
            <div
              class="col-10 col-md-6 bg-light rounded"
            >
              <h3
                class="m-0 text-center"
              >
                Main Task
              </h3>
            </div>
          </div>
          <div
            class="row g-0 mt-3 justify-content-around overflow-auto task-dependencies"
          >
            <div
              class="col-10 col-md-4"
            >
              <div
                class="list-group"
              >
                <div
                  class="bg-secondary-subtle text-secondary fw-bold text-center list-group-item"
                >
                  All columns with tasks
                </div>
                <div
                  class="list-group-item"
                >
                  <span
                    class="mb-3 badge bg-primary"
                  >
                    In Progress
                  </span>
                  <div>
                    <div
                      class="border border-3 rounded mb-3 list-group-item"
                    >
                      <p>
                        Implement ProductDetailPage component
                      </p>
                      <button
                        class="w-100 btn btn-outline-primary btn-sm"
                        type="button"
                      >
                        Add as a dependency
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="col-10 col-md-4"
            >
              <div
                class="mt-3 mt-md-0 list-group"
              >
                <div
                  class="bg-secondary-subtle text-secondary fw-bold text-center list-group-item"
                >
                  Dependency tasks
                </div>
                <div
                  class="list-group-item"
                >
                  <p>
                    Dependency Task
                  </p>
                  <button
                    class="w-100 btn btn-outline-danger btn-sm"
                    type="button"
                  >
                    Remove as a dependency
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row g-0 vh-100"
      style="background-image: url(geometric-triangle-shapes-background.jpg); background-size: cover;"
    >
      <div
        class="col-12"
      >
        <div
          class="row g-0 bg-light p-2 shadow align-items-center"
        >
          <div
            class="col-2"
          >
            <a
              class="btn btn-light bi bi-arrow-left border"
              href="/"
            />
          </div>
          <div
            class="col-8"
          >
            <h3
              class="m-0 text-center"
            >
              Task Dependencies
            </h3>
          </div>
        </div>
        <div
          class="row g-0 mt-3 justify-content-center"
        >
          <div
            class="col-10 col-md-6 bg-light rounded"
          >
            <h3
              class="m-0 text-center"
            >
              Main Task
            </h3>
          </div>
        </div>
        <div
          class="row g-0 mt-3 justify-content-around overflow-auto task-dependencies"
        >
          <div
            class="col-10 col-md-4"
          >
            <div
              class="list-group"
            >
              <div
                class="bg-secondary-subtle text-secondary fw-bold text-center list-group-item"
              >
                All columns with tasks
              </div>
              <div
                class="list-group-item"
              >
                <span
                  class="mb-3 badge bg-primary"
                >
                  In Progress
                </span>
                <div>
                  <div
                    class="border border-3 rounded mb-3 list-group-item"
                  >
                    <p>
                      Implement ProductDetailPage component
                    </p>
                    <button
                      class="w-100 btn btn-outline-primary btn-sm"
                      type="button"
                    >
                      Add as a dependency
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="col-10 col-md-4"
          >
            <div
              class="mt-3 mt-md-0 list-group"
            >
              <div
                class="bg-secondary-subtle text-secondary fw-bold text-center list-group-item"
              >
                Dependency tasks
              </div>
              <div
                class="list-group-item"
              >
                <p>
                  Dependency Task
                </p>
                <button
                  class="w-100 btn btn-outline-danger btn-sm"
                  type="button"
                >
                  Remove as a dependency
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
